src/main/java/gift/service/ProductService.java

```diff
@@ -56,4 +59,14 @@ public void deleteProduct(Long productId) {
         }
         productRepository.deleteById(productId);
     }
+
+    public List<ProductResponseDto> findAllProducts() {
+        List<Product> products = productRepository.findAllProducts();
+        List<ProductResponseDto> result = new ArrayList<>();
+        for (Product product : products) {
+            result.add(product.toDto());
+        }
+        return result;
+    }
```

Comment:
전혀 잘못 짠 코드는 아니지만, 요렇게 바꾸는 방식도 있다는걸 알려드리기 위해서 코멘트 남깁니다 :)
```suggestion
public List<ProductResponseDto> findAllProducts() {
    return productRepository.findAllProducts()
            .stream()
            .map(Product::toDto)
            .collect(Collectors.toList());
}
```
관심있다면, 람다식과 Stream에 대해서도 한 번 찾아봐도 좋을거에요.
이런 것도 있다 정도로만 알고 넘어가셔도 됩니다!

Reply:
개념서로 공부할 때는 이렇게 쓴다~로만 알고 있었는데 다시 공부해볼 기회가 되었습니다. 감사합니다!

## 람다

> 메서드를 하나의 식으로 표현한 것
> 

```java
int addOperation(int a, int b) {
return a + b;
}
->(int a,int b) -> a + b
->(a, b) -> a * b //더 간단히 가능!

from (a) -> a * a to a -> a * a //parameter 괄호 생략도 가능
(String name, int i) -> System.out.println(name + "=" + i) //중괄호와 세미콜론 생략도 가능
```

### so, What is 람다식?

- 람다식을 변수로 추출 시 Runnable @Functionalnterface가 나왔다.

### what is function interface?

- 추상 메서드가 단 하나만 있는 interface
(*default method나 static method는 존재할 수 있다.)

### 인스턴스를 만들려면?

- 추상메서드를 overriding해서 class를 만들고 이 class로부터 instance를 만들 수 있었다.

```java
@FunctionalInterface
public interface BridgeNumberGenerator {
int generate();
}

BridgeNumberGenerator bridgeRandomNumberGenerator = () -> new Random().nextInt();
int randomNumber = bridgeRandomNumberGenerator.generate();

int generate();
() -> new Random().nextInt();
추상메서드와 람다식이 1:1
```

**즉, 람다식은 함수형 인터페이스를 구현한 인스턴스**

- 함수형 인터페이스는 남이 만든 것을 사용하자.
Runnable void run() return : x // 즉 매개변수와 반환값이 모두 없어서 Runnable을 쓰는구나~로 이해 가능
Supplier<T> T get() O(T)
Consumer<T> void accept(T t) return : x
Function<T,R> R apply(T t) return : O(R)
Predicate<T> boolean test(T t) return : O(boolean)
모두 java.util.function에 속하며 reference :

### 왜 위의 것들을 사용하는 것이 좋은가?

- default method를 제공해서 다른 코드와 '상호운용성'이 뛰어나다고 함.
- Predicate에서 and() or() negate()로 조건을 조합해 하나의 새로운 Predicate로 결합 가능하다??

```java
public interface Predicate<T> {
boolean test(T t);
default Predicate<T> or(Predicate<? super T> other) {
Objects.requireNonNull(other);
return (t) -> test(t) || other.test(t);
}
```

### lambda 식을 안쓰면 결합이 불가했던것인가?

- 아무리 생각해봐도 위 코드를 람다 없이 어떻게 짜야할 지 몰라서 gpt 사용..
```java
return new Predicate<T>() {
    @Override
    public boolean test(T t) {
        return this.test(t) || other.test(t);
    }
};
```
내부 this를 사용하여 무한루프에 걸리냐 외부 인터페이스의 test method냐의 차이였다!

### ?는 뭐더라 와일드카드인 건 알겠는데 T이거나..

-  T 하위거나

### 람다식 간략화 하기

(player -> player.getName()) 은 (player::getName) 으로 바꿀 수 있다.

```java
(player -> player.getName()) // (player.getNmae)

```

---

## 스트림

## what is stream.

- 항상 사전을 찾아보자.
    
    1.개울 2. 물줄기 3. 계속 이어진 줄
    즉, 무언가 줄줄이 **이어져있는** CS에서는 **연속된 정보를 처리하기** 위해 사용된다.
    +(Modern java in action에서는 data 처리 연산을 지원하도록 source에서 추출된 연속된 요소라고 한다..)
    
    ### whyy?
    
    - 과거의 그 데이터의 형태가 무엇이었던 같은 방식으로 다룰 수 있게 되어 코드의 재사용성이 높아진다.
    
    ```java
    String[] strArr = {"aa", "bb", "cc", "dd"};
    List<String> strList = Array.asList(strArr);
    Stream<String> strStream1 = strList.stream();
    Stream<String> strStream2 = Array.stream(strArr);
    는
    strStream1.sorted().forEach(System.out::println);
    strStream2.sorted().forEach(System.out::println);
    으로 스트림 생성 방식이 달라도 데이터 처리 방법은 같을 수 있다.
    ```
    

### 스트림은 생성,중간연산,최종연산으로 이루어져있다.

```java
List<String> names = crews.stream()
.filter(crew -> crew.getCourse().equals(Course.BACKEND))
.map(Crew::getName)
.limit(10)
.collect(toList());
// .stream() -> 중간연산들 -> .collect() 최종연산
일단 스트림을 생성만하면 이후 연산과정은 동일하기에 스트림 생성이 중요하겠군.
```

1. Collection으로 부터
.stream() 으로 스트림 생성
2.  Array와 List는 
static method로 .stream()
3.  숫자로부터
.ints(); .longs(), .doubles(); 로 가능
4.  람다식으로부터
iterate() //seed값과 람다식 (seed값은 파라미터에 들어가는 첫번째 값을 의미)
나
generate()를 사용하자.
seed는 없고 lamdba로 계산

### 생성 알았으니 중간연산

1. distinct() : stream에서 중복된 data 제거, 중복 존재시 throws exception
2. filter() : 조건을 만족하는 data만 남기고 나머지는 제외

```java
public static Menu findMenuByName(String Name) {
return menus.stream()
.filter(element -> element.getName().equals(name)
//element의 이름이 실제로 같은지 체크하고 실제로 같은 데이터만을 남기고 나머지는 제외하는 함수라고 할 수 있따.
```

1. sorted(Comparator<T> comparator)
스트림을 지정한 comparator로 정렬하고
아무것도 안 들어가있을 시 기본정렬기준으로!
2. map은 원하는 필드만 뽑아내거나 특정 형태로 반환하는 method

```java
this.players = names.stream() // Stream<String>
.map(Name::new) //Stream<Name>
.map(name -> new Player(name, cards.giveInitialCards()))
.collect(Collectors.toList()); //Stream<Player>
```

```java
private static Deque<Card> initialzeCards() {
Deque<Card> cards = new ArrayDeque<>();
for (Value value : Value.values()) {
for (Shape shape : Shape.values()) {
cards.push(new Card(value, shape));
}
}
return cards;
} //음.. 싫다 싫어
```

```java
private static Deque<Card> initializeCards() {
List<Card> cards = Arrays.stream(Value.values())
.map(value -> Arrays.stream(Shape.values())
.map(shape -> new Card(value, shape)))
//그러나 Stream<Card>가 아닌 Stream<Stream<Card>>로 오류
.collect(Collectors.toList());
return new ArrayDeque<>(cards);
}
//따라서 Shape.values())에서 flatMap을 사용하자.
//평탄화 작업이 필요! 스트림의 원소인 values들을 또 매핑을 하면 여러개의 스트림이 생기게 될텐데 다시 하나의 스트림으로 합쳐주는 작업.
```

### 마지막 최종 연산!

최종 연산은 말 그대로 최종이기 때문에 한 번 수행되고
여러번 수행하고 싶다면 스트림을 여러번 생성해야한다.

1. forEach(Consumer<? super T> action)
스트림의 데이터를 소모하여 출력하는 용도
allMatch / anyMatch / nonMatch로 Predicate를 받아 요소들의 '조건 검사'도 가능하다.
    
    ```java
    return cards.stream() //Stream<Card>
    .map(Card::getNumber) // Stream<CardNumber>
    .anyMatch(cardNumber -> cardNumber.equals(CardNumber.ACE));
    //가지고 있는 카드 중에 에이스 카드가 있는가
    // 근데 이거 중간연산에서 수행할 수 있지 않은가?
    ```
    
2. 또 findFirst() // 조건에 일치하는 첫 번째 요소 반환
findAny() // 조건에 일치하는 요소를 하나 반환
3. 또 collect(Collector collector)
스트림의 요소를 수집하여 원하는 형태로 반환한다.
파라미터 Collector는 how에 관한 것..
용어 한 번 짚고가자.
`collect() vs Collector vs Collectors`

> 전자는 스트림의 최종연산으로 컬렉터를 param으로 받고
Collector은 interface로 매우 많ㅇ는 추상 메서드를 가진다.
Collectors는 매번 다 정의할 수 없으니 이미 정의된 Collector를 static method로 제공하는 클래스. 메서드가 매우 많아~
그 중 엄선
.collect(Collectors.toList()) 또는 Collectors.toMap(keyMapper, valueMapper)
> 

```java
Map<String, Student> studentsMap =
students.stream().collect(
toMap(student -> student.getName(),
student -> student));
// 또 문제 있다. -> 중복 키! -> 3번째 BinaryOperator로 (existing,replacement) -> existing)을 해줘야 우리가 생각하는 Map으로 동작
```

1. 또 groupingBy() : 스트림의 데이터들을 카테고리별 그룹화

```java
Map<Country, List<Person>> peopleByCountry =
people.stream().collect(
groupingBy(person -> person.getCountry()));
//분류함수를 인자로 받아서 데이터를 그룹화.
```

```java
Map<String,Long> freq =
words.collect(
groupingBy(word -> word.toLowerCase(),
Counting()));
```

1.  또 joining() : 스트림의 요소들을 묶어서 반환

```java
.collect(Collectors.joining(", "));
```

1. 또 reduce() : 데이터를 줄여나가며 연산을 수행하고 최종 결과 반환

```java
IntStream.ranceClosed(1, 5).reduce(0,(number1,number2) -> number1 + number2); // 스트림 요소를 줄여가며 합을 계산하는
//굳이 reduce가 아니더라도
//IntStream,DoubleStream,LongStream의을 통해
cards.Stream.mapToInt(card -> card.getNumber().getScore())
.sum(); // int를 Integer로 다시 만들어주는 과정이 한 번 있다.
```

### 그래서 stream은 언제 사용해야하나

- 반복문 : return/break/continue 또는 지역변수 수정해야할 때는 for!
그것이 아니라 연속된 데이터에서 하나의 연산으로 합칠 때나 필터링이나 일관되게 무언가를 할 때는 stream!